### PR TITLE
Log in to Wire with an SSO Link

### PR DESCRIPTION
**Pull request by makingthematrix**
If the user clicks on a link of a special type (`wire://start-sso/wire-[SSO token]`) the app opens and the user is taken directly to the Okta login page, just as if he or she typed SSO credentials in the Company Login popup.

Different than in the case of Company Login, using the link makes it possible to try to login when all slots for accounts have been used up. This case is also handled. Also, the token is consumed on opening the Okta page, so if the user decides to go back, he or she will be able to continue with the standard login procedure.

#### APK
[Download build #12329](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12329/artifact/build/artifact/wire-dev-PR1983-12329.apk)
[Download build #12334](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12334/artifact/build/artifact/wire-dev-PR1983-12334.apk)
[Download build #12335](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12335/artifact/build/artifact/wire-dev-PR1983-12335.apk)
[Download build #12336](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12336/artifact/build/artifact/wire-dev-PR1983-12336.apk)
[Download build #12340](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12340/artifact/build/artifact/wire-dev-PR1983-12340.apk)
[Download build #12341](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12341/artifact/build/artifact/wire-dev-PR1983-12341.apk)
[Download build #12348](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12348/artifact/build/artifact/wire-dev-PR1983-12348.apk)
[Download build #12349](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12349/artifact/build/artifact/wire-dev-PR1983-12349.apk)
[Download build #12351](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12351/artifact/build/artifact/wire-dev-PR1983-12351.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 1]
```diff
@@ -72,16 +69,20 @@ trait SSOFragment extends FragmentHelper {
...
-  protected def extractTokenAndShowSSODialog(showIfNoToken: Boolean = false): Unit = {
-    if (findChildFragment[InputDialog](SSODialogTag).isEmpty) {
-      extractTokenFromClipboard
-        .filter(_.nonEmpty || showIfNoToken)
-        .foreach(showSSODialog)(Threading.Ui)
-    }
+  protected def extractTokenAndShowSSODialog(showIfNoToken: Boolean = false): Unit =
+    userAccountsController.ssoToken.foreach {
+      case Some(token) =>
+        userAccountsController.ssoToken.mutate(_ => None)
```
Comment (JPonte) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: Why not just publish?

File: app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala [hunk 2]
```diff
@@ -39,13 +39,28 @@ class UserAccountsController(implicit injector: Injector, context: Context, ec:
...
+  private lazy val zms             = inject[Signal[ZMessaging]]
+  private lazy val accountsService = inject[AccountsService]
+  private lazy val prefs           = inject[Signal[UserPreferences]]
+  private lazy val convCtrl        = inject[ConversationController]
 
-  val accounts = accountsService.accountManagers.map(_.toSeq.sortBy(acc => (acc.teamId.isDefined, acc.userId.str)))
-  val convCtrl = inject[ConversationController]
+  lazy val accounts = accountsService.accountManagers.map(_.toSeq.sortBy(acc => (acc.teamId.isDefined, acc.userId.str)))
+
+  private val numberOfLoggedInAccounts = Signal(0)
```
Comment (JPonte) [inline app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala]: Why is this a `Signal` if you never listen to it? Seems to be just a `var` of `Int`.

File: app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala [hunk 3]
```diff
@@ -39,13 +39,28 @@ class UserAccountsController(implicit injector: Injector, context: Context, ec:
...
+
+  val onAllLoggedOut = EventStream[Unit]()
```
Comment (JPonte) [inline app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala]: Will an event stream be enough for cases where the event context wasn't on?

File: app/src/main/scala/com/waz/zclient/MainActivity.scala [hunk 4]
```diff
@@ -165,73 +161,100 @@ class MainActivity extends BaseActivity
     Option(ZMessaging.currentGlobal).foreach(_.googleApi.checkGooglePlayServicesAvailable(this))
   }
 
+  private def openSignUpPage(ssoToken: Option[String]): Unit = {
+    userAccountsController.ssoToken.mutate(_ => ssoToken)
```
[the remark this answers is not in the document]
Comment (makingthematrix) [reply app/src/main/scala/com/waz/zclient/MainActivity.scala]: Changed to publish

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala (hunk 1 above)
Comment (makingthematrix) [reply app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: Changed to publish

File: app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala (hunk 2 above)
Comment (makingthematrix) [reply app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala]: Changed to var

File: app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala (hunk 3 above)
Comment (makingthematrix) [reply app/src/main/scala/com/waz/zclient/common/controllers/UserAccountsController.scala]: Changed to a signal

**Pull request merged**